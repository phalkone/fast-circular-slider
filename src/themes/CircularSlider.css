.container {
  position:relative;
  stroke: #91d5ff;
  width: fit-content;
  height: fit-content;
}

.container:hover {
  stroke: #69c0ff;
}

.handle {
  fill: white;
  stroke-width: 2;
}

.selected {
  fill:#69c0ff;
  opacity:0.2;
}

.circle {
  fill: transparent;
  stroke-width: 4;
}

.backgroundCircle {
  stroke: #dedede;
}

.arc {
  fill: none;
  stroke-width: 4;
}

.degreeField {
  position: absolute;
}

.degree {
  display: inline-block;
  position: relative;
}

.degree::after {
  position: absolute;
  top: 2px;
  right: 1px;
  content: 'Â°';
}

.input {
  width: 30px;
  height: 20px;
  padding:1px 2px;
  text-align: left;
  text-indent: 1px;
  background-color: transparent;
  border: 1px solid #d9d9d9;
  border-radius: 2px;
  font-size:14px;
  box-sizing: content-box;
}

.selectedDegree,
.input:focus {
  border: 1px solid #69c0ff;
}

.input:focus {
  outline: none;
  box-shadow: 0 0 0 2px rgb(24 144 255 / 20%)
}

.input::-webkit-outer-spin-button,
.input::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.input[type=number] {
  -moz-appearance: textfield;
}
