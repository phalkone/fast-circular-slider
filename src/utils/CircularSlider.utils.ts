export const getXY = (angle : number, radius: number, padding: number) : {
  x: number,
  y : number
} => {
  const a = toRad(angle)
  const x = radius + padding + (Math.sin(a) * radius)
  const y = radius + padding - (Math.cos(a) * radius)
  return { x, y }
}

export const toDeg = (angle : number) : number => angle * (180 / Math.PI)

export const toRad = (angle : number) : number => angle * (Math.PI / 180)

export const bound = (angle : number) : number => {
  if (angle >= 360) {
    return angle - 360
  } else if (angle < 0) {
    return angle + 360
  } else {
    return angle
  }
}

export const getAngle = (mouseX : number, mouseY : number, center : {
  x: number,
  y: number
}) : number => {
  const a = Math.atan((mouseX - center.x) / (center.y - mouseY))
  let angle = Math.round(toDeg(a))
  if (mouseY <= center.y) {
    angle = angle < 0 ? 360 + angle : angle
  } else {
    angle = 180 + angle
  }
  return angle
}
