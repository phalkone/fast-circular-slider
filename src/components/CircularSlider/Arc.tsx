import React from 'react'
import type { IArc, IMousePos, IInitialArc } from '../../types/CircularSlider.types'
import { getAngle, bound } from '../../utils/CircularSlider.utils'
import { Draggable } from './Draggable'

/**
 * Arc for the slider
 */
export const Arc = (props: IArc) => {
  const onMouseDown = (parent : IMousePos) : IInitialArc => {
    const initialCenter = {
      x: parent.x + props.center,
      y: parent.y + props.center
    }
    return {
      ...initialCenter,
      start: props.start,
      end: props.end,
      arc: getAngle(parent.pageX, parent.pageY, initialCenter)
    }
  }

  const onMouseMove = (x: number, y: number, initial: IInitialArc) : void => {
    if (initial.arc && initial.start && initial.end) {
      const newAngle = getAngle(x, y, { x: initial.x, y: initial.y })
      const diff = newAngle - initial.arc
      const start = bound(initial.start + diff)
      const end = bound(initial.end + diff)
      props.setStart(start)
      props.setEnd(end)
    }
  }

  return (
    <Draggable
    onMouseDown={onMouseDown}
    onMouseMove={onMouseMove}
    initial={{ x: 0, y: 0, start: 0, end: 0, arc: 0 }}
    onTop={false}>
      <path
        d={`M ${props.startPoint.x} ${props.startPoint.y} ` +
        `A ${props.radius} ${props.radius} 0 ${props.largeFlag} 0 ` +
        `${props.endPoint.x} ${props.endPoint.y}`}
        fill='none'
        stroke='#69c0ff'
        strokeWidth='4'
      />
    </Draggable>
  )
}
