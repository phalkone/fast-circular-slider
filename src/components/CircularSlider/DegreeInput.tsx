import React, { useState } from 'react'
import type { ChangeEvent } from 'react'
import type { IDegreeInput } from '../../types/CircularSlider.types'
import * as styles from '../../themes/CircularSlider.css'

/**
 * Input field for degrees
 */
export const DegreeInput = (props: IDegreeInput) => {
  const setDegree = (e: ChangeEvent<HTMLInputElement>) => {
    let degree = +e.target.value
    if (degree > 360) degree = 360
    if (degree < 0) degree = 0
    props.setValue(props.id === 1 ? [props.value[0], degree] : [degree, props.value[1]])
  }
  const selected = (props.selectedHandle === props.id || props.selectedHandle === 2)
  const [focus, setFocus] = useState<boolean>(false)
  const onFocus = () => { setFocus(true) }
  const onBlur = () => { setFocus(false) }

  return (
    <div
      className={styles.degreeField}
      style= {{
        top: `${props.center + props.topOffset}px`,
        left: `${props.center - 25}px`
      }}
    >
      <div className={styles.degree}>
        <input
          type='number'
          maxLength={3}
          style={{
            border: (selected || focus) ? `1px solid ${props.style.hoverColor}` : `1px solid ${props.style.circleColor}`,
            boxShadow: focus ? `0 0 0 2px ${props.style.hoverColor}33` : 'none'
          }}
          className={styles.input}
          value={props.value[props.id].toString().padStart(3, '0')}
          onFocus={onFocus}
          onBlur={onBlur}
          onChange={setDegree}
          disabled={props.style.disabled}
          title={props.id === 0 ? 'Starting angle' : 'Ending angle'}
        />
      </div>
    </div>
  )
}
