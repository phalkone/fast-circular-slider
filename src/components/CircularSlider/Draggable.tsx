import React, { useState, useEffect } from 'react'
import type { IDraggable, IPosition, IInitialArc } from '../../types/CircularSlider.types'
import { getPosition } from '../../utils/CircularSlider.utils'

/**
 * To make component draggable
 */
export const Draggable = (props: IDraggable) => {
  const [initial, setInitial] = useState<IPosition | IInitialArc>({
    x: 0, y: 0, start: 0, end: 0, arc: 0
  })
  const [dragging, setDragging] = useState<boolean>(false)

  const onMouseDown = (e : React.MouseEvent<SVGElement> | React.TouchEvent<SVGElement>) : void => {
    const page = getPosition(e)
    if (e instanceof MouseEvent && e.button !== 0) return
    setDragging(true)
    const target = e.target as SVGElement
    if (target.parentNode) {
      const parent : DOMRect = (target.parentNode.parentNode as SVGElement).getBoundingClientRect()
      setInitial(props.onMouseDown({
        x: parent.x,
        y: parent.y,
        pageX: page.x,
        pageY: page.y
      }))
    }
    e.stopPropagation()
    e.preventDefault()
  }

  const onMouseUp : { (e: MouseEvent| TouchEvent): void } = (e : MouseEvent | TouchEvent) => {
    setDragging(false)
    document.removeEventListener('mousemove', onMouseMove)
    document.removeEventListener('mouseup', onMouseUp)
    document.removeEventListener('touchmove', onMouseMove)
    document.removeEventListener('touchend', onMouseUp)
    e.stopPropagation()
    e.preventDefault()
  }

  const onMouseMove : { (e: MouseEvent | TouchEvent): void } = (e : MouseEvent | TouchEvent) => {
    if (!dragging) return
    const page = getPosition(e)
    props.onMouseMove(page.x, page.y, initial)
    e.stopPropagation()
    e.preventDefault()
  }

  useEffect(() => {
    if (dragging) {
      document.addEventListener('mousemove', onMouseMove)
      document.addEventListener('mouseup', onMouseUp)
      document.addEventListener('touchmove', onMouseMove)
      document.addEventListener('touchend', onMouseUp)
    }
    if (props.onDrag) props.onDrag(dragging)
  }, [dragging])

  return (
    <g
      onMouseDown={onMouseDown}
      onTouchStart={onMouseDown}
      {...(dragging && props.onTop ? { id: props.onTop } : {})}
      >
        {props.children}
    </g>
  )
}
