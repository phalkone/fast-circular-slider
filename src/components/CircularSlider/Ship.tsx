import React from 'react'
import { IShip } from '../../types/CircularSlider.types'

/**
 * Outline of the ship
 */
export const Ship = (props: IShip) => {
  return (
    <path
      d={`M ${props.center + 10} ${props.center - 31} ` +
         `C ${props.center + 10} ${props.center - 35} ${props.center + 7} ` +
         `${props.center - 51} ${props.center} ${props.center - 51} ` +
         `C ${props.center - 7} ${props.center - 51} ${props.center - 10} ` +
         `${props.center - 35} ${props.center - 10} ${props.center - 31} ` +
         `L ${props.center - 10} ${props.center + 23} ` +
         `C ${props.center - 10} ${props.center + 33} ${props.center - 4} ` +
         `${props.center + 51} ${props.center - 4} ${props.center + 51} ` +
         `L ${props.center + 5} ${props.center + 51} ` +
         `C ${props.center + 5} ${props.center + 51} ${props.center + 10} ` +
         `${props.center + 33} ${props.center + 10} ${props.center + 23} ` +
         `L ${props.center + 10} ${props.center - 31} Z`}
      fill='none'
      stroke={props.circleColor}
      strokeWidth='4'
    />
  )
}
