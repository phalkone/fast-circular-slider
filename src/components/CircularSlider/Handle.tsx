import React from 'react'
import type { IHandle, IPosition, IMousePos } from '../../types/CircularSlider.types'
import { getAngle } from '../../utils/CircularSlider.utils'
import { Draggable } from './Draggable'

/**
 * Handle for the slider
 */
export const Handle = (props: IHandle) => {
  const onMouseEnter = () => props.setHover(true)
  const onMouseLeave = () => props.setHover(false)
  const onMouseDown = (parent : IMousePos) : IPosition => {
    return {
      x: parent.x + props.center,
      y: parent.y + props.center
    }
  }

  const onMouseMove = (x: number, y: number, initial: IPosition) : void => {
    const angle : number = getAngle(x, y, initial)
    props.setValue(props.id === 1 ? [props.value[0], angle] : [angle, props.value[1]])
  }

  const onDrag = (dragging: boolean) : void => {
    props.setSelectedHandle(dragging ? props.id : 3)
  }

  let handle = (
    <circle
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
      cx={props.points[props.id].x}
      cy={props.points[props.id].y}
      r={props.style.handleRadius}
      fill='white'
      data-testid={props.id === 0 ? 'startHandle' : 'endHandle'}
      strokeWidth={props.style.handleWidth}
      stroke={props.style.disabled ? props.style.disabledColor : (props.hover || props.selectedHandle !== 3) ? props.style.hoverColor : props.style.sliderColor}
    />
  )

  if (!props.style.disabled) {
    handle = (
      <Draggable
        onMouseDown={onMouseDown}
        onMouseMove={onMouseMove}
        onTop={props.onTop}
        onDrag={onDrag} >
        {props.selectedHandle === props.id &&
          <circle
            cx={props.points[props.id].x}
            cy={props.points[props.id].y}
            r={props.style.handleRadius + props.style.handleWidth + 2}
            fill={`${props.style.hoverColor}`}
            opacity={0.2}
          />}
          {handle}
      </Draggable>
    )
  }

  return handle
}
