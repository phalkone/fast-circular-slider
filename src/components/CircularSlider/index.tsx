import React, { useState } from 'react'
import type { ICircularSlider } from '../../types/CircularSlider.types'
import { getXY } from '../../utils/CircularSlider.utils'
import * as styles from '../../themes/CircularSlider.css'
import { Handle } from './Handle'
import { Arc } from './Arc'
import { DegreeInput } from './DegreeInput'

/**
 * Draws a circular slider for provided relative wind data
 * @param {ICircularSlider} props Properties of the CircularSlider as descibed by ICircularSlider.
 */
const CircularSlider = (props: ICircularSlider) => {
  const center = props.radius + props.padding
  const [start, setStart] = useState(props.defaultStart)
  const [end, setEnd] = useState(props.defaultEnd)

  return (
    <div className={styles.container}>
      <DegreeInput
        leftOffset={-40}
        value={start.toString().padStart(3, '0')}
        setAngle={setStart}
        center={center}
      />
      <DegreeInput
        leftOffset={3}
        value={end.toString().padStart(3, '0')}
        setAngle={setEnd}
        center={center}
      />
      <svg
        version='1.1'
        width={center * 2}
        height={center * 2}
        id='circular-slider'
      >
        <circle
          cx={center}
          cy={center}
          r={props.radius}
          stroke="#dedede"
          fill="transparent"
          strokeWidth="4"
        />
        {(start === end || Math.abs(start - end) === 360) &&
          <circle
            cx={center}
            cy={center}
            r={props.radius}
            stroke="#69c0ff"
            fill="transparent"
            strokeWidth="4"
          />
        }
        <Arc
          largeFlag={ (end < start ? 360 - start + end : end - start) > 180 ? 1 : 0 }
          startPoint={{ ...getXY(end, props.radius, props.padding) }}
          endPoint={{ ...getXY(start, props.radius, props.padding) }}
          start={start}
          end={end}
          setStart={setStart}
          setEnd={setEnd}
          radius={props.radius}
          center={center}
        />
        <Handle
          setAngle={setStart}
          angle={{ ...getXY(start, props.radius, props.padding) }}
          center={center}
        />
        <Handle
          setAngle={setEnd}
          angle={{ ...getXY(end, props.radius, props.padding) }}
          center={center}
        />
        <use href='#active' />
      </svg>
    </div>
  )
}

CircularSlider.defaultProps = {
  radius: 50,
  padding: 8,
  defaultStart: 320,
  defaultEnd: 40
}

export default CircularSlider
