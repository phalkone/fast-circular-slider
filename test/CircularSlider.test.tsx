import React from 'react'
import { fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import CircularSlider from '../src/components/CircularSlider'
import { getXY } from '../src/utils/CircularSlider.utils'

describe('Circular Slider', () => {
  test('renders Circular Slider component', () => {
    render(<CircularSlider
            value={[340, 60]} />)

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(340)
    expect(screen.getAllByRole('spinbutton')[1]).toHaveValue(60)
    expect(screen.getAllByRole('spinbutton')[3]).toBeUndefined()
  })

  test('number validation of input field', () => {
    render(<CircularSlider
            value={[340, 60]} />)

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(340)

    userEvent.type(screen.getAllByRole('spinbutton')[0], '{selectall}20')

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(20)

    userEvent.type(screen.getAllByRole('spinbutton')[0], '{selectall}460')

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(360)

    userEvent.type(screen.getAllByRole('spinbutton')[0], '{selectall}-20')

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(0)

    userEvent.type(screen.getAllByRole('spinbutton')[0], '{selectall}test')

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(0)

    userEvent.type(screen.getAllByRole('spinbutton')[0], '{selectall}60')

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(60)

    userEvent.type(screen.getAllByRole('spinbutton')[0], '{selectall}2^23')

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(223)
  })

  test('slider moving when value of input field', () => {
    render(<CircularSlider
            value={[340, 60]}
            circleRadius={70}
            handleRadius={6}
            handleWidth={2} />)

    userEvent.type(screen.getAllByRole('spinbutton')[0], '{selectall}20')

    expect(screen.getAllByRole('spinbutton')[0]).toHaveValue(20)

    let point = getXY(20, 70, 10)

    expect(screen.getByTestId('startHandle').getAttribute('cx')).toBe(point.x.toString())
    expect(screen.getByTestId('startHandle').getAttribute('cy')).toBe(point.y.toString())

    userEvent.type(screen.getAllByRole('spinbutton')[1], '{selectall}300')

    expect(screen.getAllByRole('spinbutton')[1]).toHaveValue(300)

    point = getXY(300, 70, 10)

    expect(screen.getByTestId('endHandle').getAttribute('cx')).toBe(point.x.toString())
    expect(screen.getByTestId('endHandle').getAttribute('cy')).toBe(point.y.toString())
  })

  test('full circle when start and end angle are the same', () => {
    render(<CircularSlider
            value={[340, 20]}
            circleRadius={70}
            handleRadius={6}
            handleWidth={2} />)
    userEvent.type(screen.getAllByRole('spinbutton')[1], '{selectall}340')
    const point = getXY(340, 70, 10)

    expect(screen.getByTestId('startHandle').getAttribute('cx')).toBe(point.x.toString())
    expect(screen.getByTestId('startHandle').getAttribute('cy')).toBe(point.y.toString())
    expect(screen.getByTestId('endHandle').getAttribute('cx')).toBe(point.x.toString())
    expect(screen.getByTestId('endHandle').getAttribute('cy')).toBe(point.y.toString())
    expect(screen.getByTestId('fullCircle')).toBeDefined()

    userEvent.type(screen.getAllByRole('spinbutton')[1], '{selectall}40')
    expect(screen.getAllByRole('spinbutton')[1]).toHaveValue(40)
    expect(screen.queryByTestId('fullCircle')).toBeNull()
  })

  test('drag handle to change value of input field', () => {
    render(<CircularSlider
            value={[340, 20]}
            circleRadius={70}
            handleRadius={6}
            handleWidth={2}
            style={{ position: 'absolute', top: 0, left: 0 }}/>)

    const point = getXY(315, 70, 10)
    const endHandle = screen.getByTestId('endHandle')

    fireEvent.mouseDown(endHandle)
    fireEvent.mouseMove(endHandle, { clientX: point.x, clientY: point.y })
    fireEvent.mouseUp(endHandle)
    expect(screen.getAllByRole('spinbutton')[1]).toHaveValue(315)
  })
})

test('Snapshot test of circular slider', () => {
  const { asFragment } = render(
    <CircularSlider
      value={[320, 40]}
    />
  )

  const frag = asFragment()
  const svg = frag.querySelector('svg')
  const use = svg?.querySelector('use')
  if (use) svg?.removeChild(use)

  expect(frag).toMatchSnapshot()
})
