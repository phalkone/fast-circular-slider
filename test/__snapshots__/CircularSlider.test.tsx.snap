// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test of circular slider 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div
        style="top: 42px; left: 53px;"
      >
        <div>
          <input
            maxlength="3"
            style="border: 1px solid #dedede; box-shadow: none;"
            title="Starting angle"
            type="number"
            value="320"
          />
        </div>
      </div>
      <div
        style="top: 84px; left: 53px;"
      >
        <div>
          <input
            maxlength="3"
            style="border: 1px solid #dedede; box-shadow: none;"
            title="Ending angle"
            type="number"
            value="040"
          />
        </div>
      </div>
      <svg
        height="156"
        width="156"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="78"
          cy="78"
          fill="none"
          r="68"
          stroke="#dedede"
          stroke-width="4"
        />
        <path
          d="M78 82L71.5 74L78 77L84.5 74L78 82Z"
          fill="#dedede"
        />
        <g>
          <path
            d="M 121.70955745868467 25.908977867909492 A 68 68 0 0 0 34.29044254131531 25.908977867909513"
            fill="none"
            stroke="#91d5ff"
            stroke-width="4"
          />
        </g>
        <g>
          <circle
            cx="34.29044254131531"
            cy="25.908977867909513"
            data-testid="startHandle"
            fill="white"
            r="6"
            stroke="#91d5ff"
            stroke-width="2"
          />
        </g>
        <g>
          <circle
            cx="121.70955745868467"
            cy="25.908977867909492"
            data-testid="endHandle"
            fill="white"
            r="6"
            stroke="#91d5ff"
            stroke-width="2"
          />
        </g>
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
